src/Crdt.Core/CommitBase.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.IO.Hashing;
+using System.Text.Json.Serialization;
+
+namespace Crdt.Core;
+
+public class CommitBase
+{
+    public const string NullParentHash = "0000";
+    [JsonConstructor]
+    protected CommitBase(Guid id, string hash, string parentHash, HybridDateTime hybridDateTime)
+    {
+        Id = id;
+        Hash = hash;
+        ParentHash = parentHash;
+        HybridDateTime = hybridDateTime;
+    }
+
+    public CommitBase(Guid id)
+    {
+        Id = id;
+        Hash = GenerateHash(NullParentHash);
+        ParentHash = NullParentHash;
+    }
+
+    public CommitBase() : this(Guid.NewGuid())
+    {
+    }
+
+    public (DateTimeOffset, long, Guid) CompareKey => (HybridDateTime.DateTime, HybridDateTime.Counter, Id);
+    public Guid Id { get; }
+    public required HybridDateTime HybridDateTime { get; init; }
+    public DateTimeOffset DateTime => HybridDateTime.DateTime;
+    [JsonIgnore]
+    public string Hash { get; private set; }
+
+    [JsonIgnore]
+    public string ParentHash { get; private set; }
+    public CommitMetadata Metadata { get; init; } = new();
+
+    public void SetParentHash(string parentHash)
+    {
+        Hash = GenerateHash(parentHash);
+        ParentHash = parentHash;
+    }
+
+    public string GenerateHash(string parentHash)
+    {
+        var idBytes = Id.ToByteArray();
+        var parentHashBytes = Convert.FromHexString(parentHash);
+        Span<byte> hashBytes = stackalloc byte[idBytes.Length + parentHashBytes.Length];
+        idBytes.AsSpan().CopyTo(hashBytes);
+        parentHashBytes.AsSpan().CopyTo(hashBytes[idBytes.Length..]);
+        return Convert.ToHexString(XxHash64.Hash(hashBytes));
+    }
+
+    public required Guid ClientId { get; init; }
+
+    public override string ToString()
+    {
+        return $"{Id} [{DateTime}]";
+    }
+}
+
+public class CommitBase<TChange> : CommitBase
```

Comment:
So keeping a non-generic CommitBase makes the querying code easier, but do we expect any client to need it? Could we make it internal to keep things cleaner?

Reply:
Actually it's used by the lexbox server so that it can just store changes as a blob of json. That said I could probably move that version of Commit into this repo and just make lexbox use that. The main reason is to keep lexbox application version agnostic. It also means lexbox could serve different applications even if it doesn't know the application change types.

Reply:
Just tried to do what we talked about (making CommitBase internal), but I forgot that you can't inherit from an internal class with a public class. So all the commit classes need to be public unfortunately. I will make their constructors internal, so if someone tries to inherit from them they can't. I also made the CommitBase classes abstract.